core/src/main/java/brooklyn/event/basic/MapConfigKey.java

```diff
@@ -125,6 +130,20 @@ protected Object applyEntryValueToMap(Entry value, Map target) {
         } else if (k instanceof String) {
             k = subKey((String)k);
         } else {
+            // supplier or other unexpected value
+            if (k instanceof Supplier) {
+                // TODO not thread-safe
+                Object mapAtRoot = target.get(this);
+                if (mapAtRoot==null) {
+                    mapAtRoot = new LinkedHashMap();
+                    target.put(this, mapAtRoot);
+                }
+                if (mapAtRoot instanceof Map) {
+                    synchronized (mapAtRoot) {
```

Comment:
Not convinced we should just synchronize on this map, without knowing anything about it. That feels dangerous and ad hoc. We can presumably rely on the map's put method to be synchronized if the map really needs to be synchronized. What do we expect this map (returned by `target.get(this)`) to be?